### ISSUE TITLE
New notation for array inputs

### ISSUE THREAD

**Issue by williampollet** [sunny/actor#39]
Hey @sunny !! 

Thanks for the amazing work :D 

I have an idea, regarding `inputs` and `outputs`. 

What do you think about accepting a new notation when expecting an array of similar objects, such as: 

```rb
    inputs :orders, type: [Orders]
```

(Similar to the GraphQL notation)

This way it will be easier to type arrays.

**Comment by sunny** [comment]
That makes a lot of sense! I’d be happy to merge a PR for that 🙏🏻

**Comment by sunny** [comment]
Actually, I think it would make more sense to just keep the plural and not have an array:

```rb
input :order, type: Order
inputs :orders, type: Order
```

Since now the array syntax for types is used as an "or":

```rb
input :age, type: [Integer, Float]
```

**Comment by sunny** [comment]
@williampollet Would you be willing to try a patch for this feature using the `array: true` syntax? 🙏🏻

**Comment by williampollet** [comment]
I'am taking it !

**Comment by sunny** [comment]
Good point. I think I’d still like the constraints here to be as light as possible and not try to handle all possible types.

One way we could do that would be to lean on another constraints/validation gem.

**Comment by viralpraxis** [comment]
> One way we could accept more types would be to lean on another constraints/validation gem.

Do you have examples of such gems in mind?

**Comment by sunny** [comment]
I was thinking of dry-validations, or ActiveRecord validations (perhaps with [type_validator](https://github.com/serradura/type_validator)). But there may be others.

**Comment by viralpraxis** [comment]
I think we could come up with some sort of global configuration option that will switch from our (default) type validation backend to gem-provided backend

dry-rb is more popular, should I try to implement a PoC for this? @sunny

**Comment by sunny** [comment]
I think it would make sense to use `dry-validations` but I wouldn’t want the gem as a dependency for people who are fine with the simpler current syntax.

I don’t know what the best course of action for that would be. Maybe this means that the glue code between `actor` and `dry-validations` lives in a different gem?

**Comment by sunny** [comment]
Yeah, makes sense. We could have for example:
- `config.validations = :default`,
- `config.validations = :rails` if you include the `service_actor-rails` gem,
- `config.validations = :dry_validation` with a new `service_actor-dry-validation` gem.

To allow that, perhaps the current validations could use more generic `before`, `after` & `around` hooks to avoid having to monkey-patch.

**Comment by viralpraxis** [comment]
Apart from dry-validations, can we add `array: true` syntax right now? I'm interested in this feature

**Comment by sunny** [comment]
> To allow that, perhaps the current validations could use more generic before, after & around hooks to avoid having to monkey-patch.

I was thinking of a way of “hooking” into actor from outside gems. Using for example  `before_call :…` that could be an interface that can allow outside gems the to add some validation behavior (or any kind of other behavior) and perhaps remove some other behavior (`skip_before_call :…`).

> Apart from dry-validations, can we add `array: true` syntax right now? I'm interested in this feature

Sure!

Note that you can do a similar check today with:

```rb
input :names,
      type: Array,
      must: {
        be_strings: -> array { array.all?(String) }
      }
```
